internal/storage/inmem/store.go

```diff
@@ -0,0 +1,328 @@
+package inmem
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/go-memdb"
+
+	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/internal/storage"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// Store implements an in-memory resource database using go-memdb.
+//
+// It can be used as a storage backend directly via the Backend type in this
+// package, but also handles reads in our Raft backend, and can be used as a
+// local cache when storing data in external systems (e.g. RDBMS, K/V stores).
+type Store struct {
+	db  *memdb.MemDB
+	pub *stream.EventPublisher
+}
+
+// NewStore creates a Store.
+//
+// You must call Run before using the store.
+func NewStore() (*Store, error) {
```

Comment:
With most of the 🧠s in this type, adding the Raft backend should mostly just be an exercise in plumbing it into the FSM + and pushing writes through [`raftApply`](https://github.com/hashicorp/consul/blob/9e93a30f4d3ca16de7468197842e6227b9cfec46/agent/consul/rpc.go#L898-L970).